src/main/java/com/snowplowanalytics/redash/RedashClient.java

```diff
@@ -0,0 +1,452 @@
+     *
+     * @param dataSource data transfer object which should contain all necessary information.
+     * @return int id of successfully (if so) created instance. In this case object that was transferred as argument
+     * receives that id.
+     * @throws IOException              if server is unavailable due to connection error or the API key is invalid.
+     * @throws IllegalArgumentException if there is already a data-source with this name on the Redash server.
+     */
+    public int createDataSource(DataSource dataSource) throws IOException, IllegalArgumentException {
+        if (isEntityAlreadyExists(getDataSources(), dataSource.getName())) {
+            throw new IllegalArgumentException(DATA_SOURCE_ALREADY_EXISTS);
+        }
+        String url = baseUrl + DATA_SOURCES_URL_PREFIX + API_KEY_URL_PARAM + apiKey;
+        String response = post(url, new Gson().toJson(dataSource), CheckResponseStatus.YES);
+        int id = getIdFromJson(response);
+        dataSource.setId(id);
+        return id;
+    }
+
+    /**
+     * Updates data-source on Redash server with provided as argument DataSource instance. Please note that all the argument's
+     * fields should be present and have correct values.
+     * @param dataSource Argument's name field is used to find specific data-source on the server which (if found) then
+     *                   will be updated by it's fields.
+     * @return false, if any of the argument's fields have null or empty value.
+     * True if entity successfully updated.
+     * @throws IOException with the message that the server response in the case if data-source with provided name
+     *                     couldn't be found, or wrong API key, or there is any connection error.
+     */
+    public boolean updateDataSource(DataSource dataSource) throws IOException {
+        DataSource fromDataBase;
+        try {
+            fromDataBase = getDataSource(dataSource.getName());
+        } catch (IllegalArgumentException e) {
+            throw new IOException(e.getMessage());
+        }
+        if (dataSourceIsInValid(dataSource)) {
+            return false;
+        }
+        String url = baseUrl + DATA_SOURCES_URL_PREFIX + "/" + fromDataBase.getId() + API_KEY_URL_PARAM + apiKey;
+        post(url, new Gson().toJson(dataSource), CheckResponseStatus.YES);
+        return true;
+    }
+
+    /**
+     * @return ArrayList<DataSource> of all existing data-sources on the server. If there are no data-sources then the list
+     * will be empty.
+     * @throws IOException if server is not available due to connection error or API key is invalid.
+     */
+    public List<DataSource> getDataSources() throws IOException {
+        String url = baseUrl + DATA_SOURCES_URL_PREFIX + API_KEY_URL_PARAM + apiKey;
+        Type listType = new TypeToken<ArrayList<DataSource>>() {}.getType();
+        return new Gson().fromJson(get(url, CheckResponseStatus.YES), listType);
```

Comment:
`DataSource` objects are not mapped as expected - all getters apart from `getType()` throw a `NullPointerException` because `Options` I'm assuming is null (when tested against one of our internal Redash instances)

Reply:
The same situation as it is with Group mapping.